src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,176 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// DiagnosticHandler notifies DiagnosticSource subscribers about outgoing Http requests
+    /// </summary>
+    internal sealed class DiagnosticsHandler : DelegatingHandler
+    {
+        /// <summary>
+        /// DiagnosticHandler constructor
+        /// </summary>
+        /// <param name="innerHandler">Inner handler: Windows or Unix implementation of HttpMessageHandler. 
+        /// Note that DiagnosticHandler is the latest in the pipeline </param>
+        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler) {}
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            //do not write to diagnostic source if request is invalid or cancelled,
```

Comment:
I would like this to be as lean as possible when logging is off.  Thus I would like it to look like (which is NOT an async method) 

if (s_diagnosticListener.IsEnabled(HttpActivityWriteName))
      return SendAyncWithLogging(request, cancelationToken);
else 
     return base.SendAsync(request, cancelationToken);

This makes it as lean as it can be when logging is off.   
In SendAsyncWithLogging, we want to put the sampling logic in and make that super lean (when you don't sample you do next to nothing).   Only after can we 'relax' a bit, and have less efficient logic